### PR TITLE
Parallel automl docs

### PR DESCRIPTION
**Pull request by chukarsten**
Fixes: #2559 

Adds some documentation to explain the new CFEngine functionality and the undocumented DaskEngine functionality.

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -279,9 +279,9 @@ def resolve_xref(self, env, fromdocname, builder, typ, target, node, contnode):
 
 def setup(app):
     p = Path("/home/docs/.ipython/profile_default/startup")
-    p.mkdir(parents=True, exist_ok=True)
-    shutil.copy("disable-warnings.py", "/home/docs/.ipython/profile_default/startup/")
-    shutil.copy("set-headers.py", "/home/docs/.ipython/profile_default/startup")
+    # p.mkdir(parents=True, exist_ok=True)
```
Comment (freddyaboulton) [inline docs/source/conf.py:282]: We need to add these back in before merge

File: docs/source/user_guide/automl.ipynb [hunk 2]
```diff
@@ -45,11 +45,32 @@
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 2,
```
Comment (freddyaboulton) [inline docs/source/user_guide/automl.ipynb:48]: Can you `Restart Kernel and Clear Output` before merge?

File: docs/source/user_guide/automl.ipynb [hunk 3]
```diff
@@ -645,17 +1598,1393 @@
...
+      "Loky-backed parallel loops cannot be called in a multiprocessing, setting n_jobs=1\n"
+     ]
+    }
+   ],
+   "source": [
+    "from dask.distributed import Client, LocalCluster\n",
+    "\n",
+    "from evalml.automl.engine import DaskEngine\n",
+    "\n",
+    "dask_engine_t2 = DaskEngine(Client(LocalCluster(processes=True, n_workers = 2)))\n",
```
Comment (freddyaboulton) [inline docs/source/user_guide/automl.ipynb:2864]: In the final cell, `t2` is listed as two threads but it's actually two processes right?

image

File: docs/source/user_guide/automl.ipynb [hunk 4]
```diff
@@ -645,17 +1598,1393 @@
...
+  },
+  {
+   "cell_type": "markdown",
    "metadata": {},
+   "source": [
+    "### Parallelism with Concurrent Futures\n",
+    "\n",
+    "The `EngineBase` class is robust and extensible enough to support futures-like implementations from a variety of libraries.  The `CFEngine` extends the `EngineBase` to use the native Python concurrent.futures library.  The `CFEngine` supports both thread- and process-level parallelism.  The type of parallelism can be chosen using either the `ThreadPoolExecutor` or the `ProcessPoolExecutor`.  If either executor is passed a `max_workers` parameter, it will set the number of processes and threads spawned.  If not, the default number of processes will be equal to the number of processors available and the number of threads set to five times the number of processors available.\n",
+    "\n",
+    "Note: the cell demonstrating process-level parallelism is commented out due to incompatibility with our ReadTheDocs build.  It can be run successfully locally."
```
Comment (freddyaboulton) [inline docs/source/user_guide/automl.ipynb:2666]: nit: I think it would be easier to understand why the code is commented out if this sentence is right above the code.

File: docs/source/user_guide/automl.ipynb [hunk 5]
```diff
@@ -645,17 +1598,1393 @@
...
+    "automl_cf_threaded = AutoMLSearch(X_train=X, y_train=y,\n",
+    "                      problem_type=\"binary\",\n",
+    "                      allowed_model_families=[ModelFamily.LINEAR_MODEL],\n",
+    "                      engine = threaded_cf_engine)\n",
+    "automl_cf_threaded.search(show_iteration_plot = False)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 28,
```
Comment (angela97lin) [inline docs/source/user_guide/automl.ipynb:2741]: I see the notebook has been run--is this necessary? Or can we clear the output?

File: docs/source/conf.py (hunk 1 above)
Comment (angela97lin) [inline docs/source/conf.py:282]: 👀 ?

File: docs/source/user_guide.ipynb [hunk 6]
```diff
@@ -100,7 +100,7 @@
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "Python 3",
+   "display_name": "Python 3 (ipykernel)",
```
Comment (angela97lin) [inline docs/source/user_guide.ipynb:103]: Same comment, is this change necessary? :o

File: docs/source/user_guide/automl.ipynb [hunk 7]
```diff
@@ -645,17 +1598,1393 @@
...
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2021-07-27T18:56:20.147287Z",
+     "start_time": "2021-07-27T18:56:20.145197Z"
+    }
+   },
    "outputs": [],
    "source": [
-    "automl.results"
+    "# from concurrent.futures import ProcessPoolExecutor\n",
```
Comment (freddyaboulton) [inline docs/source/user_guide/automl.ipynb:2750]: I think this would look nicer if we made this cell a markdown cell and put the code in a "```python" code block

**Review by freddyaboulton — APPROVED** [review]
@chukarsten Thanks for putting this together! I echo @angela97lin 's comment about clearing the output before merging. We also need to revert the changes to `conf.py`. Other than that, left some formatting nits.

Thanks again!

File: docs/source/release_notes.rst [hunk 8]
```diff
@@ -10,6 +10,7 @@ Release Notes
         * Renamed ``ComponentGraph``'s ``get_parents`` to ``get_inputs`` :pr:`2540`
         * Removed ``ComponentGraph.linearized_component_graph`` and ``ComponentGraph.from_list`` :pr:`2556`
     * Documentation Changes
+        * Added documentation for `DaskEngine` and `CFEngine` parallel engines :pr:`2560`
```
Comment (bchen1116) [inline docs/source/release_notes.rst:13]: double tic marks on `DaskEngine` and `CFEngine` rather than single?

File: docs/source/user_guide/automl.ipynb [hunk 9]
```diff
@@ -645,17 +1598,1393 @@
...
+     ]
+    }
+   ],
+   "source": [
+    "from concurrent.futures import ThreadPoolExecutor\n",
+    "\n",
+    "from evalml.automl.engine import CFEngine\n",
+    "from evalml.automl.engine.cf_engine import CFClient\n",
+    "\n",
+    "# Use thread-level paralellism\n",
```
Comment (bchen1116) [inline docs/source/user_guide/automl.ipynb:2729]: Typo: Parallelism

File: docs/source/user_guide/automl.ipynb [hunk 10]
```diff
@@ -645,17 +1598,1393 @@
...
+    "threaded_cf_engine = CFEngine(CFClient(ThreadPoolExecutor()))\n",
+    "\n",
+    "automl_cf_threaded = AutoMLSearch(X_train=X, y_train=y,\n",
+    "                      problem_type=\"binary\",\n",
+    "                      allowed_model_families=[ModelFamily.LINEAR_MODEL],\n",
+    "                      engine = threaded_cf_engine)\n",
```
Comment (bchen1116) [inline docs/source/user_guide/automl.ipynb:2735]: Hypernit: Can we fix the spacing here? This is what I see
image

File: docs/source/user_guide/automl.ipynb [hunk 11]
```diff
@@ -645,17 +1598,1393 @@
...
+    "# automl_cf_process.search(show_iteration_plot = False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Parallelism with Dask\n",
+    "\n",
+    "Thread or process level parallelism can be explicitly evoked for the DaskEngine (as well as the CFEngine).  The `processes` can be set to `True` and the number of processes set using `n_workers`.  If `processes` is set to `False`, then the resulting parallelism will be threaded and `n_workers` will represent the threads used.  Examples of both will follow."
```
Comment (bchen1116) [inline docs/source/user_guide/automl.ipynb:2768]: Another hypernit: Can we add tic marks around `DaskEngine` and `CFEngine` here? It was done previously and would be nice to keep the pattern
image

File: docs/source/user_guide/automl.ipynb [hunk 12]
```diff
@@ -645,17 +1598,1393 @@
...
+   "source": [
+    "from dask.distributed import Client, LocalCluster\n",
+    "\n",
+    "from evalml.automl.engine import DaskEngine\n",
+    "\n",
+    "dask_engine_t2 = DaskEngine(Client(LocalCluster(processes=True, n_workers = 2)))\n",
+    "automl_dask_t2 = AutoMLSearch(X_train=X, y_train=y,\n",
+    "                      problem_type=\"binary\",\n",
+    "                      allowed_model_families=[ModelFamily.LINEAR_MODEL],\n",
+    "                      engine = dask_engine_t2)\n",
```
Comment (bchen1116) [inline docs/source/user_guide/automl.ipynb:2868]: nit with spacing here as well, but not as bad as the previous!

File: docs/source/user_guide/automl.ipynb [hunk 13]
```diff
@@ -645,17 +1598,1393 @@
...
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/usr/local/Caskroom/miniconda/base/envs/evalml_dev/lib/python3.9/site-packages/distributed/node.py:160: UserWarning:\n",
+      "\n",
+      "Port 8787 is already in use.\n",
+      "Perhaps you already have a cluster running?\n",
+      "Hosting the HTTP server on port 50005 instead\n",
```
Comment (bchen1116) [inline docs/source/user_guide/automl.ipynb:2890]: This might just be a warning message you got locally, but would be nice to clear and make sure
image

File: docs/source/user_guide/automl.ipynb [hunk 14]
```diff
@@ -645,17 +1598,1393 @@
...
+      "Concurrent futures (threaded) search duration: 3.979705810546875\n",
+      "Dask (two threads) search duration: 9.009714126586914\n",
+      "Dask (four processes)search duration: 4.219135046005249\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"Sequential search duration: %s\" % str(automl.search_duration))\n",
+    "print(\"Concurrent futures (threaded) search duration: %s\" % str(automl_cf_threaded.search_duration))\n",
+    "# print(\"Concurrent futures (process) search duration: %s\" % str(automl_cf_process.search_duration))\n",
```
Comment (bchen1116) [inline docs/source/user_guide/automl.ipynb:2979]: nit: It kinda looks awk to have this one commented out. I think you can remove since you've gotten the point across regardless

**Review by bchen1116 — APPROVED** [review]
Left nits aplenty, but once the local run is cleared, should be much better!

File: docs/source/user_guide/automl.ipynb (hunk 13 above)
Comment (freddyaboulton) [reply docs/source/user_guide/automl.ipynb:2890]: My hope is that the warning is supressed when the changes to `conf.py` are reverted.

File: docs/source/user_guide/automl.ipynb (hunk 5 above)
Comment (chukarsten) [reply docs/source/user_guide/automl.ipynb:2741]: I don't understand how its possible I submitted this unclear.  I make it a habit to clear all output before I add and commit.

File: docs/source/user_guide/automl.ipynb (hunk 2 above)
Comment (chukarsten) [reply docs/source/user_guide/automl.ipynb:48]: The problem lies in me forgetting to hit save one last time before adding and committing :P. I'm still not used to ipynb in the codebase.

File: docs/source/user_guide/automl.ipynb (hunk 3 above)
Comment (chukarsten) [reply docs/source/user_guide/automl.ipynb:2864]: Good eye

File: docs/source/user_guide/automl.ipynb (hunk 4 above)
Comment (chukarsten) [reply docs/source/user_guide/automl.ipynb:2666]: @freddyaboulton thoughts?

image

File: docs/source/user_guide/automl.ipynb (hunk 7 above)
Comment (chukarsten) [reply docs/source/user_guide/automl.ipynb:2750]: see the comment in the previous?

File: docs/source/user_guide/automl.ipynb (hunk 4 above)
Comment (freddyaboulton) [reply docs/source/user_guide/automl.ipynb:2666]: Looks good!

File: docs/source/release_notes.rst (hunk 8 above)
Comment (chukarsten) [reply docs/source/release_notes.rst:13]: ngl, I have no idea what the difference is :)

File: docs/source/user_guide/automl.ipynb (hunk 13 above)
Comment (chukarsten) [reply docs/source/user_guide/automl.ipynb:2890]: Yea,apparently this comes from not putting the LocalCluster instantiation inside a `if __name__ == "__main__":` block.  I think this will go away with #2561

File: docs/source/user_guide/automl.ipynb [hunk 15]
```diff
@@ -651,11 +651,131 @@
...
+    "from evalml.automl.engine.cf_engine import CFClient\n",
+    "\n",
+    "# Use thread-level parallelism\n",
+    "threaded_cf_engine = CFEngine(CFClient(ThreadPoolExecutor()))\n",
+    "\n",
+    "automl_cf_threaded = AutoMLSearch(X_train=X, y_train=y,\n",
+    "                                  problem_type=\"binary\",\n",
+    "                                  allowed_model_families=[ModelFamily.LINEAR_MODEL],\n",
+    "                                  engine = threaded_cf_engine)\n",
+    "automl_cf_threaded.search(show_iteration_plot = False)"
```
Comment (eccabay) [inline docs/source/user_guide/automl.ipynb:691]: nit: remove the spaces around `=` here

Comment (eccabay) [reply docs/source/user_guide/automl.ipynb:691]: And a few lines down as well!

File: docs/source/user_guide/automl.ipynb [hunk 16]
```diff
@@ -651,11 +651,131 @@
...
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Parallelism with Dask\n",
+    "\n",
+    "Thread or process level parallelism can be explicitly evoked for the `DaskEngine` (as well as the `CFEngine`).  The `processes` can be set to `True` and the number of processes set using `n_workers`.  If `processes` is set to `False`, then the resulting parallelism will be threaded and `n_workers` will represent the threads used.  Examples of both will follow."
```
Comment (eccabay) [inline docs/source/user_guide/automl.ipynb:719]: nit: "Examples of both will follow" --> "Examples of both follow"

File: docs/source/user_guide/automl.ipynb [hunk 17]
```diff
@@ -651,11 +651,131 @@
...
+    "from dask.distributed import Client, LocalCluster\n",
+    "\n",
+    "from evalml.automl.engine import DaskEngine\n",
+    "\n",
+    "dask_engine_p2 = DaskEngine(Client(LocalCluster(processes=True, n_workers = 2)))\n",
+    "automl_dask_p2 = AutoMLSearch(X_train=X, y_train=y,\n",
+    "                              problem_type=\"binary\",\n",
+    "                              allowed_model_families=[ModelFamily.LINEAR_MODEL],\n",
+    "                              engine = dask_engine_p2)\n",
+    "automl_dask_p2.search(show_iteration_plot = False)\n",
```
Comment (eccabay) [inline docs/source/user_guide/automl.ipynb:737]: More spaces around ` = `!

File: docs/source/user_guide/automl.ipynb [hunk 18]
```diff
@@ -651,11 +651,131 @@
...
+    "                              allowed_model_families=[ModelFamily.LINEAR_MODEL],\n",
+    "                              engine = dask_engine_t4)\n",
+    "automl_dask_t4.search(show_iteration_plot = False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As we can see, a significant performance gain can result from simply using something other than the default `SequentialEngine`."
```
Comment (eccabay) [inline docs/source/user_guide/automl.ipynb:760]: Can you be more specific about what sort of performance gain comes from this?

**Review by eccabay — APPROVED** [review]
LGTM! Just some really small nitpicks from me

File: docs/source/user_guide/automl.ipynb (hunk 15 above)
Comment (chukarsten) [reply docs/source/user_guide/automl.ipynb:691]: thanks!

File: docs/source/user_guide/automl.ipynb (hunk 17 above)
Comment (chukarsten) [reply docs/source/user_guide/automl.ipynb:737]: just to bother you! ;)

File: docs/source/user_guide/automl.ipynb (hunk 18 above)
Comment (chukarsten) [reply docs/source/user_guide/automl.ipynb:760]: Hmmm, I'd like to, but I'm a little hesitant to specifically quantify something that can vary a little bit with different runs.

File: docs/source/user_guide/automl.ipynb [hunk 19]
```diff
@@ -651,11 +651,131 @@
...
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Parallelism with Dask\n",
+    "\n",
+    "Thread or process level parallelism can be explicitly evoked for the `DaskEngine` (as well as the `CFEngine`).  The `processes` can be set to `True` and the number of processes set using `n_workers`.  If `processes` is set to `False`, then the resulting parallelism will be threaded and `n_workers` will represent the threads used.  Examples of both follow."
```
Comment (ParthivNaresh) [inline docs/source/user_guide/automl.ipynb:719]: `evoked` or `invoked`? ;)

**Review by ParthivNaresh — APPROVED** [review]
Clean and concise!
